client/service/src/client/client.rs

```diff
@@ -817,18 +818,6 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 					}
 				};
 
-				// ensure parent block is finalized to maintain invariant that
-				// finality is called sequentially.
-				if finalized {
```

Comment:
Why remove this? Maybe guard it under `parent_exists` instead?

Reply:
This is a weird hack. Why only finalized a single parent? What if their parent is not finalized as well?
Sequential finalization is already insured by the DB layer. There's `ensure_sequential_finalization` function that is called on block commit. I believe this code was added before `ensure_sequential_finalization` and is no longer needed.

Reply:
`apply_finality_with_block_hash` was taking the tree route from the last finalized block and making sure all intermediary blocks are finalized, but yeah you're right that this is guaranteed by the DB layer now. The only thing extra that `apply_finality_with_block_hash` is doing is sending finality notifications for all these intermediary blocks (although we cap it to 256 blocks), I am not sure if we still should keep doing that as some subsystems might be expecting these finality notifications (@rphmeier is this needed for parachains?).

Reply:
That's a good point. I'll add it back with a comment.